File: python/dsti-generator/src/templates/guard_tests.py
"""Generate guard clause tests for Python."""
# @docspec:module {
#   id: "docspec-py-dsti-guard-tests",
#   name: "Guard Clause Test Template",
#   description: "Generates pytest test classes that validate guard clause enforcement, producing one test per guard plus a None-rejection test for each method.",
#   since: "3.0.0"
# }


# @docspec:method { since: "3.0.0" }
# @docspec:deterministic
# @docspec:intentional "Generate guard clause test file content from DSTI intent signals"
def generate_guard_tests(method: dict, output_dir: str) -> list[dict]:
    qualified = method.get("qualified", "unknown")
    parts = qualified.rsplit("#", 1)
    class_name = parts[0].rsplit(".", 1)[-1] if parts else "Unknown"
    method_name = parts[1] if len(parts) > 1 else "unknown_method"
    test_file = f"test_{class_name.lower()}_{method_name}_guards.py"

    signals = method.get("intentSignals", {})
    guard_clauses = signals.get("guardClauses", 0)
    guard_count = guard_clauses if isinstance(guard_clauses, int) else len(guard_clauses) if isinstance(guard_clauses, list) else 0

    tests = ""
    for i in range(guard_count):
        guard = guard_clauses[i] if isinstance(guard_clauses, list) and i < len(guard_clauses) else None
        condition = guard.get("condition", f"guard condition {i + 1}") if isinstance(guard, dict) else f"guard condition {i + 1}"
        error = guard.get("error", "ValueError") if isinstance(guard, dict) else "ValueError"

        tests += f'''
    def test_guard_{i + 1}_{method_name}(self):
        """Should enforce guard: {condition}"""
        # Given: input that violates the guard
        # When/Then: expect {error}
        with pytest.raises({error}):
            # TODO: Call {method_name} with invalid input
            pass

'''

    tests += f'''
    def test_{method_name}_rejects_none(self):
        """Should reject None input."""
        with pytest.raises((TypeError, ValueError)):
            # TODO: Call {method_name} with None
            pass
'''

    content = f'''"""Guard clause tests for {class_name}#{method_name}.

Auto-generated from DSTI intent signals.
"""
import pytest


class Test{class_name.capitalize()}{method_name.capitalize()}Guards:
{tests}
'''

    return [{"path": f"{output_dir}/{test_file}", "content": content, "type": "guard"}]

File: rust/dsti-generator/src/templates/guard_tests.rs
//! Generate guard clause test functions.
//!
//! @docspec:module {
//!   id: "docspec-rust-dsti-templates-guard",
//!   name: "Guard Clause Test Template",
//!   description: "Generates #[should_panic] test stubs for each guard clause detected in a method's DSTI intent signals, ensuring boundary conditions are tested.",
//!   since: "3.0.0"
//! }

use serde_json::Value;
use crate::generator::GeneratedTestFile;

/// @docspec:deterministic
/// @docspec:intentional "Generates one #[should_panic] test per detected guard clause from DSTI signals"
pub fn generate(method: &Value, output_dir: &str) -> Vec<GeneratedTestFile> {
    let qualified = method.get("qualified").and_then(|q| q.as_str()).unwrap_or("unknown");
    let parts: Vec<&str> = qualified.rsplitn(2, '#').collect();
    let method_name = parts.first().copied().unwrap_or("unknown");
    let type_name = parts.get(1).and_then(|s| s.rsplit("::").next()).unwrap_or("Unknown");
    let file_name = format!("test_{}_guards.rs", method_name);

    let signals = method.get("intentSignals").unwrap();
    let guard_count = signals.get("guardClauses").and_then(|g| g.as_u64()).unwrap_or(0);

    let mut tests = String::new();

    for i in 0..guard_count {
        tests.push_str(&format!(r#"
    #[test]
    #[should_panic]
    fn test_guard_{i}_{method_name}() {{
        // Given: input that violates guard condition {num}
        // When/Then: expect panic
        // TODO: Call {method_name} with invalid input
        todo!("Implement guard test");
    }}
"#, i = i, method_name = method_name, num = i + 1));
    }

    let content = format!(r#"//! Guard clause tests for {type_name}::{method_name}.
//! Auto-generated from DSTI intent signals.

#[cfg(test)]
mod {method_name}_guard_tests {{
    use super::*;
{tests}
}}
"#, type_name = type_name, method_name = method_name, tests = tests);

    vec![GeneratedTestFile {
        path: format!("{}/{}", output_dir, file_name),
        content,
        test_type: "guard".to_string(),
    }]
}
